util/argo/ref_sources.go

```diff
@@ -0,0 +1,62 @@
+package argo
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	argoappv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/db"
+)
+
+type GetRefSourcesOptions struct {
+	Sources    argoappv1.ApplicationSources
+	Db         db.ArgoDB
+	Revisions  []string
+	IsRollback bool
+}
+
+// GetRefSources creates a map of ref keys (from the sources' 'ref' fields) to information about the referenced source.
+// This function also validates the references use allowed characters and does not define the same ref key more than
+// once (which would lead to ambiguous references).
+// In case of rollback, this function also updates the targetRevision to the proper revision
+func GetRefSources(ctx context.Context, opts GetRefSourcesOptions) (argoappv1.RefTargetRevisionMapping, error) {
+	refSources := make(argoappv1.RefTargetRevisionMapping)
+	if len(opts.Sources) > 1 {
```

Comment:
We can avoid the check `len(opts.Sources) > 1` as the loop would not be executed if `opts.Sources` is empty.

Reply:
Yeah, but you still can have 1 single item in sources slides, that's the reason. Originally this func received the app itself and checked `if spec.HasMultipleSources()` and this is the closer option.

I can remove it if you still think that it's not required